scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,104 +11,402 @@ import scalafix.ScalafixConfig
+        case tn: m.Type.Name => tn
+      }
+
+      // Mix local scope with root scope for FQN and non-FQN lookups
+      val wholeScope = mixScopes(inScope, realRootScope)
+      val disambiguatingSyntax = ref.syntax
+      val (_, reversedSymbols) = {
+        names.iterator.foldLeft(wholeScope -> List.empty[g.Symbol]) {
+          case ((scope, symbols), metaName) =>
+            val sym = lookupBothNames(metaName.value,
+                                      scope,
+                                      symbols.headOption,
+                                      disambiguatingSyntax)
+            if (!sym.exists) scope -> symbols
+            else sym.info.members -> (sym :: symbols)
+        }
+      }
+
+      val symbols = reversedSymbols.reverse
+      val metaToSymbols = names.zip(symbols)
+      logger.elem(metaToSymbols)
+
+      if (symbols.nonEmpty) {
+        /* Check for path dependent types:
+         * 1. Locate the term among the FQN
+         * 2. If it exists, get the first value in the chain */
+        val maybePathDependentType = metaToSymbols
+          .find(ms => ms._2.isAccessor)
+          .flatMap(_ =>
+            metaToSymbols.dropWhile(ms => !ms._2.isValue).headOption)
+        val isPathDependent = maybePathDependentType.isDefined
+
+        val (lastName, lastSymbol) = maybePathDependentType
+          .getOrElse(metaToSymbols.last)
+        val (onlyPaths, shortenedNames) =
+          metaToSymbols.span(_._1 != lastName)
+
+        val renames: Map[m.Name, g.Name] = shortenedNames.map {
+          case (metaName, symbol) =>
+            metaName -> st.renames.getOrElse(symbol, symbol).name
+        }.toMap
+
+        val localSym = inScope.lookup(lastSymbol.name)
+        if (lastSymbol.exists &&
+            (isPathDependent || localSym.exists)) {
+          // Return shortened type for names already in scope
+          val onlyNames = onlyPaths.map(_._1)
+          val removed = removePrefixes(refNoThis, onlyNames)
+          val shortenedAndRenamedType = removed.transform {
+            case name: m.Name =>
+              val maybeGname: Option[g.Name] = renames.get(name)
+              if (maybeGname.isDefined) {
+                val gName = maybeGname.get
+                val renamedName = {
+                  if (gName.isTypeName) m.Type.Name(gName.decoded)
+                  else m.Term.Name(gName.decoded)
+                }
+                if (renamedName.value != name.value) renamedName
+                else name
+              } else name
+          }
+          Right(shortenedAndRenamedType.asInstanceOf[m.Ref])
+        } else {
+          // Remove unnecessary packages from type name
+          val noRedundantPaths = {
+            val paths = onlyPaths.dropWhile(path =>
+              getUnderlyingPkgSymbol(path._2) != enclosingTerm)
+            if (paths.isEmpty) onlyPaths.map(_._1)
+            else paths.tail.map(_._1)
+          }
+
+          // Shortened names must be just one if no PDT
```

Comment:
What's the idiomatic way of failing in scalafix? Here we're failing fast since these cases should never happen. I wonder if this is ok or we should go for another solution.